// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.processor";
option java_outer_classname = "ProcessorProto";
option objc_class_prefix = "HLW";

package processor;

service Processor {
  rpc Ping (Empty) returns (Empty) {}
  rpc Follow (FollowRequest) returns (Empty) {}
  rpc Lead (LeadRequest) returns (Empty) {}
  rpc Prompt (PromptRequest) returns (Empty) {}
  rpc GetNotifications (NotificationsRequest) returns (NotificationsReply) {}
  rpc CallApplicationMethod (CallRequest) returns (CallReply) {}

}

message Empty {
}

message InitRequest {
  string application_name = 1;
}

message FollowRequest {
  string upstream_name = 1;
  string upstream_address = 2;
}

message LeadRequest {
  string downstream_name = 1;
  string downstream_address = 2;
}

message PromptRequest {
  string upstream_name = 1;
}

message NotificationsRequest {
  string section_id = 1;
}

message NotificationsReply {
  string section = 1;
}

message CallRequest {
  string method_name = 1;
  string args = 2;
  string kwargs = 3;
}

message CallReply {
  string data = 1;
}
